File: algorithms/javaalgo/src/test/java/leetcode/number1to100/number41to50/number42/TrappingRainWaterTest.java
package leetcode.number1to100.number41to50.number42;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;

public class TrappingRainWaterTest {

  @Test
  void test() {
    int[] height1 = {0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};
    int[] height2 = {4, 2, 0, 3, 2, 5};

    assertThat(trap(height1)).isEqualTo(6);
    assertThat(trap(height2)).isEqualTo(9);
  }

  public int trap(int[] height) {
    int answer = 0;

    int length = height.length;
    if (length == 0) {
      return answer;
    }
    int lt = 0;
    int rt = length - 1;
    int ltMax = height[lt];
    int rtMax = height[rt];

    while (lt < rt) {
      ltMax = Math.max(ltMax, height[lt]);
      rtMax = Math.max(rtMax, height[rt]);

      if (ltMax <= rtMax) {
        answer += ltMax - height[lt];
        lt++;
      } else {
        answer += rtMax - height[rt];
        rt--;
      }
    }
    return answer;
  }
}

File: algorithms/javascriptalgo/leetcode/number1-100/number41-50/number42/trappingRainWater.test.js
test("test", () => {
  const height1 = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1];
  const height2 = [4, 2, 0, 3, 2, 5];

  expect(trap(height1)).toEqual(6);
  expect(trap(height2)).toEqual(9);
});

function trap(height) {
  let answer = 0;

  let length = height.length;
  if (length === 0) {
    return answer;
  }

  let left = 0;
  let right = length - 1;
  let leftMax = height[left];
  let rightMax = height[right];

  while (left < right) {
    leftMax = Math.max(leftMax, height[left]);
    rightMax = Math.max(rightMax, height[right]);

    if (leftMax <= rightMax) {
      answer += leftMax - height[left];
      left++;
    } else {
      answer += rightMax - height[right];
      right--;
    }
  }

  return answer;
}
